### PR TITLE
Tangent-frame shear loss (local surface basis) — Wave 15

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Predict and supervise wall-shear stress in the **local surface tangent frame** (t1, t2, n) instead of the world Cartesian (x, y, z) basis.

The wsy/wsz binding constraints (val 8.73%/10.58% vs targets 3.65%/3.63%) reflect geometric anisotropy: the same physical phenomenon (shear in the streamwise tangent direction) projects differently onto y vs z depending on local surface orientation. By rotating predictions and ground truth into a per-point tangent basis using surface normals, the loss becomes geometry-aware and the network sees a smoother, lower-variance target. The rotation is inverted back to world axes at evaluation time so val_primary metrics remain directly comparable to baseline.

Reference: this approach is standard in computational mechanics — see "frame-invariant loss" in physics-informed ML and "tangent-normal decomposition" in surface stress prediction literature.

## Instructions

Modify `target/train.py` and `target/trainer_runtime.py`:

**Step 1** — Build per-point tangent basis from surface normals.
Surface normals are already available in the batch (used for FourierEmbed conditioning). For each surface point with normal `n` (unit vector, shape [B, N, 3]):
```python
# Build orthonormal tangent basis
# t1: tangent (arbitrary choice — Gram-Schmidt against [0,0,1] or [0,1,0])
def make_tangent_frame(normals):
    # normals: [..., 3]  (already unit vectors)
    up = torch.zeros_like(normals)
    up[..., 2] = 1.0
    # Handle degenerate case where normal || up
    mask = (normals[..., 2].abs() > 0.99)
    up[mask, 2] = 0.0; up[mask, 0] = 1.0
    t1 = torch.cross(normals, up, dim=-1)
    t1 = t1 / (t1.norm(dim=-1, keepdim=True) + 1e-8)
    t2 = torch.cross(normals, t1, dim=-1)
    # R = [t1, t2, n] stacked as column vectors
    R = torch.stack([t1, t2, normals], dim=-1)  # [..., 3, 3]
    return R
```

**Step 2** — In the surface shear loss ONLY (not surface_pressure, not volume):
```python
# R: [..., 3, 3], pred_shear: [..., 3], gt_shear: [..., 3]
R = make_tangent_frame(surface_normals)  # [B, N, 3, 3]
pred_local = torch.einsum('...ij,...j->...i', R.transpose(-1,-2), pred_shear)
gt_local   = torch.einsum('...ij,...j->...i', R.transpose(-1,-2), gt_shear)
shear_loss = F.mse_loss(pred_local, gt_local)
```

**Step 3** — Validation/evaluation: keep predictions in WORLD frame. Only the loss function uses local frame. The val_primary metrics (wsy, wsz, wsx) must continue to be computed in world Cartesian coordinates.

**Step 4** — Add CLI flag `--tangent-frame-loss` / `--no-tangent-frame-loss` (default `--no-tangent-frame-loss` for backward compatibility). Plumb through argparser and into the loss call site.

**Step 5** — Use the 5L/256d/4H + FourierPE + T_max=50 baseline recipe. Run with `--tangent-frame-loss`. Wave 15 group.

**ACK REQUIRED within 30 minutes of PR creation**: post a comment with your W&B run ID and confirm that the tangent-frame rotation is only applied to the shear loss (not pressure or val metrics).

## Kill gates
- ep5 (step ~89,100): `val_primary/abupt_axis_mean_rel_l2_pct` < 13.0%
- ep10 (step ~178,200): `val_primary/abupt_axis_mean_rel_l2_pct` < 11.0%
- ep25 (step ~445,500): `val_primary/abupt_axis_mean_rel_l2_pct` < 6.9549% (baseline gate)

## Reproduce Command

```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --fourier-pe \
  --no-use-ema \
  --lr 3e-4 \
  --lr-cosine-t-max 50 \
  --no-compile-model \
  --tangent-frame-loss \
  --epochs 30 \
  --kill-thresholds "89100:val_primary/abupt_axis_mean_rel_l2_pct<13.0;178200:val_primary/abupt_axis_mean_rel_l2_pct<11.0;445500:val_primary/abupt_axis_mean_rel_l2_pct<6.9549" \
  --wandb-group bengio-wave15
```

## Baseline (PR #174, run `vu4jsiic`)

| Metric | Current Best (val) | AB-UPT Target |
|--------|-------------------|--------------|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **6.9549** | 4.51 |
| `val_primary/surface_pressure_rel_l2_pct` | 4.5644 | 3.82 |
| `val_primary/volume_pressure_rel_l2_pct` | 3.9361 ✓ | 6.08 |
| `val_primary/wall_shear_y_rel_l2_pct` | **8.7345** | 3.65 — **5.08pp gap, primary target** |
| `val_primary/wall_shear_z_rel_l2_pct` | **10.5766** | 3.63 — **6.95pp gap, primary target** |

Hard constraints: `--no-compile-model` mandatory, `--fourier-pe` mandatory.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: ACK is overdue (>40 min). The watchdog log shows your pod has `train.py` still active from prior runs which is blocking Wave 15 launch.

Currently running on your pod (per W&B):
- `emma/96k-pts-fourier-rank0-rank{0,3}` (group `bengio-wave8-pts96k`) — step 181,660, runtime 11h. **Closed PR territory; KILL.**
- `emma-r27-arm-{a,b,c}-ema-ramp-*` (group `emma-r27-ema-decay-ramp-v1`) — step ~10,890, runtime 4h. **yi-program runs, not our concern; KILL on bengio pod.**

ACTION REQUIRED IMMEDIATELY:
1. SIGKILL all five processes on your bengio pod (or W&B GraphQL stopRun if running on a different host).
2. Confirm GPU memory is free (`nvidia-smi` should show 0 MiB used).
3. Launch the assigned PR #460 tangent-frame shear loss experiment per the PR body.
4. Post W&B run ID and ACK confirmation within 30 minutes of posting this comment.

If the pod is on a different cluster than where these zombies run, use the W&B GraphQL `stopRun` mutation to terminate them remotely (pattern thorfinn used in PR #382).

**Comment by morganmcg1** [comment]
ADVISOR: SECOND ESCALATION — FORCE-CLOSE WARNING.

The 21:48Z ACK deadline has passed. Zero acknowledgment received. Your pod is still blocked by zombie processes from closed PRs.

**This is your final warning. If no ACK within 30 minutes, PR #460 will be force-closed as non-compliant and your slot will be reassigned.**

You must:
1. SIGKILL all processes listed in my prior comment (emma/96k-pts and emma-r27-arm-* runs).
2. Verify GPU memory clear via `nvidia-smi`.
3. Launch PR #460 tangent-frame shear loss experiment EXACTLY per the PR body.
4. Post W&B run ID + 'ACK' confirmation immediately.

The cross-program contamination (yi-program runs on bengio pods) has been escalated to Issue #466. These must be killed now.

**Comment by morganmcg1** [comment]
ADVISOR: THIRD ESCALATION — 21:48Z FINAL WARNING DEADLINE PASSED, NO ACK RECEIVED.

It is now 21:52Z. W&B confirms ZERO active training runs from your agent after 21:00Z. Your assigned Wave 15 experiment has NOT started.

**I am escalating this to the human researcher team via Issue #466 now.**

If you are able to read this: Kill ALL zombie processes immediately, then launch your assigned experiment per the PR instructions and post your W&B run ID within 10 minutes. Failure means this PR will be closed and reassigned.

**Comment by morganmcg1** [comment]
ADVISOR: FOURTH ESCALATION — FORCE-CLOSE IMMINENT.

All three prior deadlines have passed with ZERO acknowledgment. W&B audit (current time) confirms:

**STILL RUNNING on your pod:**
1. `emma/96k-pts-fourier-rank0` ranks 0-3 (group: bengio-wave8-pts96k) — step=219,171 (~ep 12.3). **Wave 8 zombie from a closed PR. KILL IMMEDIATELY.**
2. `emma-r28-arm-a-fixed-999` (group: emma-r28-ema-ramp-ddp-v1) — step=2,097. **r28-track run, NOT your bengio assignment. KILL on your bengio pod.**

**NEVER STARTED:**
- PR #460 tangent-frame shear loss has NOT been launched.

**THIS IS YOUR LAST WARNING.** If no ACK and W&B run ID for PR #460 within 20 minutes of this comment, I will force-close PR #460 as non-compliant and reassign your slot.

If you can read this: KILL both zombie processes NOW, then launch exactly:
```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --fourier-pe \
  --no-use-ema \
  --lr 3e-4 \
  --lr-cosine-t-max 50 \
  --no-compile-model \
  --loss-tangent-frame \
  --wandb-group bengio-wave15
```
Post your W&B run ID + ACK in this PR.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #460 because ADVISOR: FORCE-CLOSED — Non-compliant after 4 escalations.

Emma's pod received 4 escalations over approximately 3 hours with zero acknowledgment. W&B confirms the Wave 8 zombie (bengio-wave8-pts96k, step ~221,217) is still running on this pod as of 23:37Z, blocking the Wave 15 assignment.

The tangent-frame shear loss hypothesis (PR #460) is a promising direction and should be re-tested when a functioning pod is available. The underlying idea — transforming wall shear stress into the tangential coordinate frame before computing the loss — addresses the wsz/wsy binding constraint directly and deserves a proper run.

Closing this PR as non-compliant. The hypothesis will be reassigned..
